### PR TITLE
Implement adminBro page for broadcast notifications

### PR DESCRIPTION
**Pull request by mohammadranjbarz**
related to #780

### CODE REVIEW

**Comment by mohammadranjbarz** [comment]
It uses this notificationType

https://github.com/Giveth/notification-center/blob/e040e6baf5f19b2df0b7fe5d5b4710d02aa30780/migrations/1660716115917-seedNotificationType.ts#L100-L132

**Comment by mohammadranjbarz** [comment]
You can create broadcast notifications from this page

Screen Shot 1401-10-13 at 17 03 27

File: src/adapters/notifications/MockNotificationAdapter.ts [hunk 1]
```diff
@@ -201,4 +204,12 @@ export class MockNotificationAdapter implements NotificationAdapterInterface {
     });
     return Promise.resolve(undefined);
   }
+
+  broadcastNotification(
+    params: BroadCastNotificationInputParams,
+  ): Promise<void> {
+    // throw new Error('test adminBro')
```
Comment (CarlosQ96) [inline src/adapters/notifications/MockNotificationAdapter.ts]: Remove comment

File: src/adapters/notifications/NotificationCenterAdapter.ts [hunk 2]
```diff
@@ -819,4 +692,259 @@ export class NotificationCenterAdapter implements NotificationAdapterInterface {
...
+    while (!allUserFetched) {
+      const { users } = await findAllUsers({ take, skip });
+      if (users.length === 0) {
+        allUserFetched = true;
+        break;
+      }
+      skip += users.length;
+      const queueData: SendBatchNotificationBody = [];
+      for (const user of users) {
+        queueData.push({
```
Comment (CarlosQ96) [inline src/adapters/notifications/NotificationCenterAdapter.ts:713]: I don't think this is necessary here @mohammadranjbarz. We can offload all the work to the notification center, because this would spam the notification center with many requests, it might hit a rate limit (not sure if we have with nginx).

Remember the notification center also have all users registered with their wallet address. 

Simply send a request with the type: "Broadcast" and the data, and handle the segment loop in the notification center in multiple queues. (I think we would need to implement the queue system in notification center that we don't currently have).

For the notifications to insert in the DB, I think we can insert all the values in one single query in the notification center.
https://stackoverflow.com/questions/24344783/insert-into-for-all-users-who-are-in-another-table

So to summarize:
1. Send 1 request with the type of notification **broadcast** and the data. (this enqueues the segment events enqueuer looper).
2. Run the query to insert all values to all users in a single query.  (I can help with this)
3. The segment events enqueuer, loops and enqueues each individual user notification to be sent with segment (in its own queue).

Its a lot more work, but its more efficient.

Comment (CarlosQ96) [reply src/adapters/notifications/NotificationCenterAdapter.ts:713]: BUT!!!!!!!!!! ***************************

If we are short in time, we can use this to start, but the loop do it inside a Queue not in the main thread. Because We had issues before where doing heavy lifting in an endpoint (like not paginating) and the backend slows down.

(this is my limited knowledge in js, for instance since it's single threaded does bull spawn new threads? or uses the same one?) (if its the same thread then doing the loop in a queue won't do anything performance wise, and you can ignore this last suggestion and merge).

**Review by CarlosQ96 — CHANGES_REQUESTED** [review]
Hey Mohammad I think the code is good and works!!!

 I have some important suggestions for performance in the loop for sending requests to the notification center. Let me know what you think and we can decide the best move forward.

File: src/adapters/notifications/NotificationCenterAdapter.ts (hunk 2 above)
Comment (mohammadranjbarz) [reply src/adapters/notifications/NotificationCenterAdapter.ts:713]: Thanks @CarlosQ96  for your comments, I remember you had mentioned on the specs of issue as well

During the holidays I and Amin talked about that and we came to this conclusion:
1. We don't have all of the users in notification-center, we just have the user's info that we already sent them at least 1 notification
2. We don't have emails of users there
3. We implemented https://staging.notification.giveth.io/docs/#/Notification/SendBulkNotification s we can send 100 notifications in one request, so if all of our users are 2,000 we can send notification for them with only 20 requests  and also we queue the requests so it would not make lots of load on notification-center
4. probably don't use this feature too much, maybe just a few times a month (maybe less) so I think a minimal implementation can be enough for that and it would not make lots of load on impact-graph

So I think yeah current implementation might be good at the beginning, do you agree with me?

File: src/adapters/notifications/MockNotificationAdapter.ts (hunk 1 above)
Comment (mohammadranjbarz) [reply src/adapters/notifications/MockNotificationAdapter.ts]: Done

**Comment by mohammadranjbarz** [comment]
@CarlosQ96 
I talked with @MoeNick  and we have to do something before merging so I made this Pr draft:
1. Add email of admin that sent notification
2. instead of sending multiple fields just send one markdown
3. preview the markdown in admin bro if it's possible

I would work on these 3 items

**Comment by mohammadranjbarz** [comment]
Screen Shot 1401-10-18 at 16 27 38
Now we changed it to get markdown as input and we show the preview of it in admin-bro and convert it to HTML and send it to notificaiton-center

@CarlosQ96

**Comment by mohammadranjbarz** [comment]
@CarlosQ96  It's ready for review now

File: migration/1672836674875-createTestAdminUser.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+                INSERT INTO public.user (email, "walletAddress", role,"loginType", name,"encryptedPassword") 
+                VALUES('test-admin@giveth.io', '0x66F014F90cdE78130e143b00F8Bd3F77c73bb0Bc', 'admin','wallet', 'test', '${hash}');
+              `);
+  }
+
+  async down(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`
+               DELETE FROM public.user where email='test-admin@giveth.io'`);
+  }
+}
```
Comment (mohammadranjbarz) [inline migration/1672836674875-createTestAdminUser.ts:25]: I added this migration file to create a test admin user when someone run migrations on local machine and wants to contribute

Comment (CarlosQ96) [reply migration/1672836674875-createTestAdminUser.ts:25]: Perfect

File: src/server/adminBro.ts [hunk 4]
```diff
@@ -162,23 +169,12 @@ export const getAdminBroRouter = async () => {
...
-          }
-          return false;
-        } catch (e) {
-          logger.error({ e });
+      authenticate: async (email, password): Promise<User | boolean> => {
+        const isPasswordCorrect = await checkAdminPassword({ email, password });
+        if (!isPasswordCorrect) {
           return false;
         }
+        return (await findAdminUserByEmail(email)) as User;
```
Comment (CarlosQ96) [inline src/server/adminBro.ts]: While I agree with the refactor, we are doing the same query twice. We should only query the user once.
Maybe the checkAdminPassword can return the user and we return it here.

**Review by CarlosQ96 — CHANGES_REQUESTED** [review]
Great work Mohammad! I think we can improve the auth handler a bit. Ill push this change and Approve

**Review by CarlosQ96 — APPROVED** [review]
Approved. There seems to be a network issue tho. Once that is fixed we can merge. improved auth handler
